kv/memberlist/dnsprovider.go

```diff
@@ -13,3 +15,31 @@ type DNSProvider interface {
 	// Addresses returns the latest addresses present in the DNSProvider.
 	Addresses() []string
 }
+
+type dnsProvider struct {
+	sync.Mutex
+	addr []string
+}
+
+func NewDNSProvider() DNSProvider { return &dnsProvider{} }
+
+func (d *dnsProvider) Resolve(ctx context.Context, addrs []string) error {
+	d.Lock()
+	defer d.Unlock()
+	for _, a := range addrs {
+		ips, err := net.LookupIP(a)
```

Comment:
Should the context be checked for cancellation on each iteration? That seems like it'd make this more responsive.

Reply:
Or perhaps just once before doing any DNS lookups?